### PR TITLE
Bitcoin block validation

### PR DESCRIPTION
**Pull request by MaxFomenkov**
Added Bitcoin block validation: 

- hash validation against the header content 
- PoW validation

### CODE REVIEW

File: src/main/kotlin/io/emeraldpay/dshackle/upstream/bitcoin/BitcoinBlockValidator.kt [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        return h <= target
+    }
+
+    private fun parseTarget(bits: Long): BigInteger {
+        val size: Int = (bits shr 24).toInt() and 0xFF
+        if (size == 0) {
+            return BigInteger.ZERO
+        }
+        val bytes = ByteArray(size) { 0 }
+        bytes[0] = (bits shr 16 and 0xFFL).toByte()
```
Comment (Termina1) [inline src/main/kotlin/io/emeraldpay/dshackle/upstream/bitcoin/BitcoinBlockValidator.kt:63]: Let's add comments to this magic

Comment (MaxFomenkov) [reply src/main/kotlin/io/emeraldpay/dshackle/upstream/bitcoin/BitcoinBlockValidator.kt:66]: added! 
additionally implementation can be observed here https://github.com/bitcoinj/bitcoinj/blob/f367ffa7094ab10816a7ebda562aab55b22a4bcd/core/src/main/java/org/bitcoinj/core/Block.java#L566

File: src/main/kotlin/io/emeraldpay/dshackle/upstream/bitcoin/BitcoinBlockValidator.kt [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+            .plus(time.toLong().asUInt32())
+            .plus(Hex.decodeHex(bits).asUInt32LE())
+            .plus(nonce.toLong().asUInt32())
+
+        return headerBytes.sha256().sha256().reversedArray().toHexString().lowercase() == hash
+    }
+
+    private fun validatePoW(bits: String, hash: String): Boolean {
+        val target = readCompactForm(bits.toLong(16))
+        val h = BigInteger(hash, 16)
```
Comment (Termina1) [inline src/main/kotlin/io/emeraldpay/dshackle/upstream/bitcoin/BitcoinBlockValidator.kt:53]: We also need some heuristic for chainwork validation, also we cannot validate it 100%, there must be someway to reduce the risk of spoofing this parameter

Comment (MaxFomenkov) [reply src/main/kotlin/io/emeraldpay/dshackle/upstream/bitcoin/BitcoinBlockValidator.kt:79]: added chainwork validation:

- chain work calculation check 
- chain work target change period check

Comment (Termina1) [reply src/main/kotlin/io/emeraldpay/dshackle/upstream/bitcoin/BitcoinBlockValidator.kt:79]: We can assume that bitcoin changes its target with fixed period, and if we know one day when it changed (and have constant for it), we can calculate the next target update

**Pull request closed without merging**